software/tracksight/frontend/src/app/components/dropdown_menu.tsx

```diff
@@ -1,52 +1,63 @@
-import { useState, useEffect, React } from 'react';
+import { useState, useEffect } from 'react'; // Removed extra React import
 import { DownOutlined, SmileOutlined } from '@ant-design/icons';
-import type { MenuProps } from 'antd';
 import { Dropdown, Space } from 'antd';
-import WebSocketComponent from './web_socket';
-
 
 const DropdownMenu = (props) => {
-    const [data, setData] = useState([]) // Format is list of signals 
-    const [items, setItems] = useState<MenuProps['items']>([]);
-    const noItems: [] = [{
-        key: "1",
-        label: "No Avaliable Signals"
-    }];
-
-    // Request available signals through socket on render
-    props.socket.emit("available_signals", {"ids" : []});
+    const [items, setItems] = useState([]);
+    const [visible, setVisible] = useState(false);
+    const [selectedSignals, setSelectedSignals] = useState([]);
 
-    useEffect(() => {
-        if (data.length > 0) {
-            const res = []
-            for (let i = 0; i < data.length; i++) {
-                res.push({
-                    key: i.toString(),
-                    label: (
-                        data[i]
-                    )
-                })
-            }
-            setItems(res)
+    const handleSignalClick = (signalName) => {
+        const newSelectedSignals = [...selectedSignals];
+        const index = newSelectedSignals.indexOf(signalName);
+        if (index === -1) {
+            newSelectedSignals.push(signalName);
         } else {
-            setItems(noItems)
+            newSelectedSignals.splice(index, 1);
         }
-    }, [data]);
+        setSelectedSignals(newSelectedSignals);
+        props.setSignal(newSelectedSignals);
+    };
+
+    useEffect(() => {
+        const updatedItems = props.signals.map((signalName, index) => ({
+            key: index.toString(),
+            label: (
+                <p>
+                    <input 
+                        type="checkbox" 
+                        checked={selectedSignals.includes(signalName)}
+                        onChange={() => handleSignalClick(signalName)} // Added onChange handler
+                    />
+                    <span onClick={() => handleSignalClick(signalName)}> 
+                        {signalName}
+                    </span>
+                </p>
+            ),
+        }));
+        setItems(updatedItems);
+    }, [props.signals, selectedSignals]);
 
-    props.socket.getAvailableSignals
+    const handleVisibleChange = (flag) => {
+        setVisible(flag);
+    };
 
     return (
-        <div>
-            <WebSocketComponent socket={props.socket} setAvailableSignals={setData} />
-            <Dropdown menu={{ items }}>
+        <div className="dropdown" onMouseLeave={() => setVisible(false)}>
+            <Dropdown 
+                menu={{ items }} 
+                open={visible} 
+                onOpenChange={handleVisibleChange}
+            >
                 <a onClick={(e) => e.preventDefault()}>
```

Comment:
Nit but can you remove this a tag, and put `style={{cursor: "pointer"}`on the Space tag for the cursor effect

Reply:
Jokes, checked the ant design docs. Instead of a tag, wrap it with <Button> it seems.

Reply:
Button from the antd components

Reply:
Also instead of manually coding the opening and closing of the menu (because we want user to click into it, not hover) you can add this onto the Dropdown tag `trigger={["click"]}`